source/common/upstream/cluster_manager_impl.cc

```diff
@@ -336,28 +338,72 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
+bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& cluster) {
   // First we need to see if this new config is new or an update to an existing dynamic cluster.
   // We don't allow updates to statically configured clusters in the main configuration.
   const std::string cluster_name = cluster.name();
-  auto existing_cluster = primary_clusters_.find(cluster_name);
-  if (existing_cluster != primary_clusters_.end() &&
-      (!existing_cluster->second.added_via_api_ ||
-       existing_cluster->second.config_hash_ == MessageUtil::hash(cluster))) {
+  const auto existing_active_cluster = active_clusters_.find(cluster_name);
+  const auto existing_warming_cluster = warming_clusters_.find(cluster_name);
+  const uint64_t new_hash = MessageUtil::hash(cluster);
+  if ((existing_active_cluster != active_clusters_.end() &&
+       existing_active_cluster->second->blockUpdate(new_hash)) ||
+      (existing_warming_cluster != warming_clusters_.end() &&
+       existing_warming_cluster->second->blockUpdate(new_hash))) {
     return false;
   }
 
-  if (existing_cluster != primary_clusters_.end()) {
-    init_helper_.removeCluster(*existing_cluster->second.cluster_);
+  if (existing_active_cluster != active_clusters_.end() ||
+      existing_warming_cluster != warming_clusters_.end()) {
+    init_helper_.removeCluster(*existing_active_cluster->second->cluster_);
+    cm_stats_.cluster_modified_.inc();
+  } else {
+    cm_stats_.cluster_added_.inc();
+  }
+
+  // There are two discrete paths here depending on when we are adding/updating a cluster.
+  // 1) During initial server load we use the init manager which handles complex logic related to
+  //    primary/secondary init, static/CDS init, warming all clusters, etc.
+  // 2) After initial server load, we handle warming independently for each cluster in the warming
+  //    map.
+  // Note: It's likely possible that all warming logic could be centralized in the init manager, but
+  //       a decision was made to split the logic given how complex the init manager already is. In
+  //       the future we may decide to undergo a refactor to unify the logic but the effort/risk to
+  //       do that right now does not seem worth it given that the logic is generally pretty clean
+  //       and easy to understand.
+  const bool use_active_map =
+      init_helper_.state() != ClusterManagerInitHelper::State::AllClustersInitialized;
+  loadCluster(cluster, true, use_active_map ? active_clusters_ : warming_clusters_);
+
+  if (use_active_map) {
+    ENVOY_LOG(info, "add/update cluster {} during init", cluster_name);
+    auto& cluster_entry = active_clusters_.at(cluster_name);
+    createOrUpdateThreadLocalCluster(*cluster_entry);
+    init_helper_.addCluster(*cluster_entry->cluster_);
+  } else {
+    auto& cluster_entry = warming_clusters_.at(cluster_name);
+    ENVOY_LOG(info, "add/update cluster {} starting warming", cluster_name);
```

Comment:
I know this probably also applies to listener warming, but I'm wondering how useful warming really is beyond the brief patching around the issues that Istio needs due to not having implemented ADS. Specifically, it seems that a config can be accepted for a set of new clusters, but left in perpetually warming state. It's kind of weird to think that (1) the management server doesn't know if the new config is really active yet (except via proving with a route update and `validate_clusters`) and we don't have anyway to cleanly dump the actual active state. This not really an actionable comment, more food for thought.

Reply:
For listeners, I think warming is absolutely required for realtime updates while serving traffic. For clusters, I think this is not a real concern in production, but given the small amount of code required I don't mind supporting. 

The "infinite warming" situation is a real concern. We could potentially report status back to the management server in a more coherent way in the future if needed.